velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
   }
 }
 
+// Blocks patterns that contain character class union, intersection, or
+// difference because these are not understood by RE2 and will be parsed as a
+// different pattern than in java.util.regex.
+void ensureRegexIsConstantAndCompatible(
+    const char* functionName,
+    const VectorPtr& patternVector) {
+  if (!patternVector ||
+      patternVector->encoding() != VectorEncoding::Simple::CONSTANT) {
+    VELOX_USER_FAIL("{} requires a constant pattern.", functionName);
+  }
+  if (patternVector->isNullAt(0)) {
+    return;
+  }
+  const StringView pattern =
+      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+  // Call the factored out function to check the pattern.
+  checkForCompatiblePattern(
+      std::string(pattern.data(), pattern.size()), functionName);
+}
+
+/// Full implementation of RegexReplace found in SparkSQL only,
+/// due to semantic mismatches betweeen Spark and Presto
+/// regex_replace(string, pattern, overwrite) → string
+/// regex_replace(string, pattern, overwrite, position) → string
+///
+/// If a string has a substring that matches the given pattern, replace
+/// the match in the string wither overwrite and return the string. If
+/// optional paramter position is provided, only make replacements
+/// after that positon in the string (1 indexed).
+///
+/// If position <= 0, throw error.
+/// If position > length string, return string.
+template <typename T>
+struct RegexReplaceFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    checkForCompatiblePattern(pattern);
+
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+
+    std::string string(stringInput.data(), stringInput.size());
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
+    if (string.size()) {
+      result.resize(string.size());
+      std::memcpy(result.data(), string.data(), string.size());
+    }
+  }
+
+  void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_CHECK_LT(-1, position - 1);
+    checkForCompatiblePattern(pattern);
+
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece(stringInput.data(), stringInput.size());
+
+    const char* prefixEnd = inputStringPiece.data() +
```

Comment:
I assume position is in terms of characters, not bytes. Is this correct? Would you add some tests with non-ASCII strings?

Reply:
I still wonder about this.

Reply:
The test regexReplaceNonAsciiPosition tests the following function call

regex_replace("♫ Resume is updated¡♫ ", "¡", "!", 2) and expects the output to be "♫ Resume is updated!♫ ". When compiled and run this test passes

Reply:
@codyschierbeck I'm not sure it is sufficient to say that a test passes. If 'position' indicates number of characters, then this code seems incorrect. I might be wrong though. Please, help explain how this code works.

Reply:
@codyschierbeck I didn't see "// Adjust the position for UTF-8 by counting the code points." code earlier. I see it now and it makes sense. Maybe use existing helper functions to skip n UTF-8 characters.